### DISCUSSION TITLE
ULAB in Micropython for SAMD51 Thing Plus and TEENSY4.1

### DISCUSSION THREAD

**Discussion by LeandroFernandes88** [v923z/micropython-ulab#665] [General]
Hello everybody.
It is possible to compile ulab to use in micropython in the boards SAMD51 Thing Plus and TEENSY4.1?
I'm looking to UF2(SAMD51) and HEX (Teensy4.1) files Micropython already with ulab and I couldn't find anything.
If there is a guide teaching how to compile this build or Iif somebody has a firmware already done I would thanks a lot.

**Comment by v923z** [comment]
The thing plus should have enough flash and RAM, so I don't see why it would be impossible to compile firmware for that. Here are the `micropython` instructions https://github.com/micropython/micropython/tree/master/ports/samd, and you can find some information on compiling for other platforms here: https://github.com/v923z/micropython-ulab/. If you manage to build the firmware, could you, please, update the readme? I would also be happy to add a build script here https://github.com/v923z/micropython-builder, so that the newest firmware would be available.

**Reply by LeandroFernandes88** [reply]
Sorry, I tried to build micropython in SAMD51 with ulab (and would try for Teensy4.1 later), followed the steps of this guide and even ask the help of a AI, but I was unsuccessful.
I'm using Windows 11 with MSYS2, install python, make, gcc, and others programs but I was unable to compile the code and mount the firmware. The last try made this error. Probably due to my little experience with micropython

**Comment by v923z** [comment]
OK, but this seems to have nothing to do with `ulab` as such. You build fails with pure `micropython`. Have you tried to ask for help here? https://github.com/orgs/micropython/discussions

**Reply by LeandroFernandes88** [reply]
goodnight.
I apologize once again.
I'm now using WLS2 and some dependencies: Python3, Cmake, gcc, gcc-arm-none-eabi, git, and others.
I was able to compile the Micropython firmware (only) for Sparkfun SAMD51 Thing Plus, following the guidance on https://github.com/micropython/micropython/tree/master/ports/samd

I tried to compile now with Ulab, but I may have made a mistake. Could you help or suggest the correct procedure?

**Comment by v923z** [comment]
You shouldn't copy anything, and you shouldn't edit any files. You simply clone two repositories, as described here: https://github.com/v923z/micropython-ulab/?tab=readme-ov-file#firmware

`ulab` is just a user module, so you can follow the steps here https://docs.micropython.org/en/latest/develop/cmodules.html

**Reply by LeandroFernandes88** [reply]
I sincerely appreciate the help.

I followed your guidance and could compile the Firmware successfully with just Micropython.

They recommended me a procedure to compile ulab in micropython and it seems that I had a problem with ulab this time.

If you can take a look at the discussion, in the last post:
https://github.com/orgs/micropython/discussions/14017

I will still study the links you sent me and try to mount the firmware again.

Thank you much

![error ulab](https://github.com/v923z/micropython-ulab/assets/125308622/2b98337d-c5f3-43f3-a666-69fa77f87af4)

**Comment by v923z** [comment]
The error is caused by this https://github.com/micropython/micropython/blob/90e517862d055b8ec95a0c7067052c0ba6cd0df1/py/parsenum.c#L219-L223 

Your platform doesn't seem to support built-in complex numbers, so the function in question is not defined. I should fix that, but in the meantime, there are multiple options here. 

First, you can simply disable the whole IO module here https://github.com/v923z/micropython-ulab/blob/63dfbd178ba3b5cfbff5d656f586d7f3f3e46409/code/ulab.h#L506-L512 and here https://github.com/v923z/micropython-ulab/blob/63dfbd178ba3b5cfbff5d656f586d7f3f3e46409/code/ulab.h#L538-L544 You can also just disable the `loadtxt` function. Or you can call the correct function in https://github.com/v923z/micropython-ulab/blob/63dfbd178ba3b5cfbff5d656f586d7f3f3e46409/code/numpy/io/io.c#L242.

**Comment by v923z** [comment]
I've fixed it in https://github.com/v923z/micropython-ulab/pull/666. Can you, please, pull the changes, and try again?

**Reply by LeandroFernandes88** [reply]
Ok, I loaded ulab again. It passed the part that had previously stopped.
However, it appears there was a problem elsewhere.
Would it be necessary to uninstall and reinstall arm-none-eabi?

I'm sorry, I'm new to microcontrollers. But I promise I will study more, and I keep trying.

Any suggestion? grateful.

![error micropython-ulab](https://github.com/v923z/micropython-ulab/assets/125308622/aae0ba9b-7719-4021-adc5-a7369779c2b5)

**Reply by v923z** [reply]
You could try to get rid of your build output directory, and recompile.

**Reply by v923z** [reply]
What you can also do is compile the firmware on github. That would have the advantage that you can work with a linux system. You can look at the scripts at https://github.com/v923z/micropython-builder. 
Finally, you can also check out https://platformio.org/. You can configure your firmware there, and the rest is taken care of.

**Comment by LeandroFernandes88** [comment]
Hello again,

I apologize for all the inconvenience I have been causing on your forum and possibly on others. Because I actually asked for help in several places due to a certain urgency of time and not knowing that everyone would be so helpful in all places.

Learning a new language, such as programming, can seem really challenging, especially for those who have a native language problem.

I sincerely apologize if I commited any mistake or exposed all your brilliant work with the creation and maintenance of ulab, which is definitely essential for my project today.

I thank everyone, who even in small collaborations, helped me get a little closer to a consistent answer.

I followed these steps, briefly:
I used WLS2, created a folder to do the work and ran the following commands:
$ sudo apt-get install gcc-arm-none-eabi (among many others)
$ git clone https://github.com/micropython/micropython.git
$ git clone https://github.com/v923z/micropython-ulab.git ulab
$ cd Micropython
$ cd mpy-cross
$ make
$ cd..
$ cd ports/samd
$ make BOARD=SPARKFUN_SAMD51_THING_PLUS submodules
$ make BOARD=SPARKFUN_SAMD51_THING_PLUS clean
$ make BOARD=SPARKFUN_SAMD51_THING_PLUS USER_C_MODULES=../../../ulab all

And unfortunately it presented the same result, there being a problem when finishing compiling the file.

I was able to successfully perform the same procedure for ports/mimxrt BOARD=TEENSY41 and it generated the necessary files, including .hex. I confirmed by uploading the firmware to the microcontroller and it seemed to work normally.

So, if you could, please take another look at this matter.
I would greatly appreciate it, from my heart.


NOTE: Would it be too abusive to ask for a Micropython UF2 file with Ulab for the SPARKFUN_SAMD51_THING_PLUS microcontroller? (PS: I don't think I would have the know-how, nor the capacity, to try other compilers at this point, after so many trys)

![error SAMD51 micropython-ulab](https://github.com/v923z/micropython-ulab/assets/125308622/a4676b19-7c01-418e-bf84-c48a510ff09a)

**Reply by LeandroFernandes88** [reply]
I tried ADAFRUIT_FEATHER_M0_EXPRESS and the same problem occurred, at the same point.

![error SAMD21 micropython-ulab](https://github.com/v923z/micropython-ulab/assets/125308622/eabac858-81d2-41a2-9840-7f14e1151ce0)

**Reply by v923z** [reply]
The error message actually tells you what the problem is: that port doesn't support complex numbers. As I pointed out earlier, you can disable complexes in `ulab.h` by setting the `ULAB_SUPPORTS_COMPLEX` constant to 0. If you do that, the firmware will compile.

**Reply by v923z** [reply]
You can download the firmware from here: https://github.com/v923z/micropython-builder/releases/tag/latest.

**Reply by LeandroFernandes88** [reply]
I truly appreciate your response, all the effort and attention you spent, without which I would not be able to continue with my work.
Thank you very much.
